python/simulation.py

```diff
@@ -0,0 +1,783 @@
+        vec2 = py_v3_to_vec(self.dims, v2)
+
+        self.swigobj = mp.volume(vec1, vec2)
+
+
+class Harminv(object):
+
+    def __init__(self, sim, c, pt, fcen, df, mxbands=None):
+        self.c = c
+        self.pt = pt
+        self.fcen = fcen
+        self.df = df
+        self.mxbands = mxbands
+        self.harminv_data = []
+        self.harminv_data_dt = 0
+        self.harminv_results = []
+        self.harminv_spectral_density = 1.1
+        self.harminv_Q_thresh = 50.0
+        self.harminv_rel_err_thresh = 1e20
+        self.harminv_err_thresh = 0.01
+        self.harminv_rel_amp_thresh = -1.0
+        self.harminv_amp_thresh = -1.0
+        self.simulation = sim
+
+    def __call__(self):
+        return self._harminv()
+
+    def _display_run_data(self, data_name, data):
+        data_str = [str(f) for f in data]
+        print("{}{}:, {}".format(data_name, self.simulation.run_index, ', '.join(data_str)))
+
+    def _collect_harminv(self, data, data_dt):
+
+        closure = {'data': data, 'data_dt': data_dt}
+
+        def _collect1(c, pt):
+            closure['data'] = []
+
+            closure2 = {'t0': 0}
+
+            def _collect2(sim):
+                closure['data_dt'] = self.simulation.meep_time() - closure2['t0']
+                closure2['t0'] = self.simulation.meep_time()
+                self.harminv_data.append(self.simulation._get_field_point(c, pt))
+            return _collect2
+        return _collect1
+
+    def _analyze_harminv(self, data, fcen, df, maxbands, dt=None):
+        self._display_run_data('harminv', ['frequency', 'imag.', 'freq.', 'Q', '|amp|', 'amplitude', 'error'])
+
+        bands = mp.py_do_harminv(self.harminv_data, dt if dt else self.simulation.fields.dt,
+                                 fcen - df / 2, fcen + df / 2, maxbands, self.harminv_spectral_density,
+                                 self.harminv_Q_thresh, self.harminv_rel_err_thresh, self.harminv_err_thresh,
+                                 self.harminv_rel_amp_thresh, self.harminv_amp_thresh)
+
+        for freq, amp, err in bands:
+            Q = freq.real / (-2 * freq.imag)
+            self._display_run_data('harminv', [freq.real, freq.imag, Q, abs(amp), amp, err])
+
+        return bands
+
+    def _harminv(self):
+        closure = {
+            '_data': [],
+            '_dt': 0,
+            '_c': self.c,
+            '_pt': self.pt,
+            '_fcen': self.fcen,
+            '_df': self.df,
+            '_maxbands': self.mxbands,
+            'data': self.harminv_data,
+            'dt': self.harminv_data_dt,
+            'results': self.harminv_results,
+        }
+
+        def _harm(sim):
+            closure['data'] = closure['_data']
+            closure['dt'] = closure['_dt']
+            if closure['_maxbands'] is None or closure['_maxbands'] == 0:
+                mb = 100
+            else:
+                mb = closure['_maxbands']
+
+            closure['results'].extend(
```

Comment:
`extend` seems wrong here... shouldn't the new results replace any old ones?

Reply:
Fixed.